### PR TITLE
Make the source a regular class

### PR DESCRIPTION
**Pull request by tobim**
###  :notebook_with_decorative_cover: Description

This introduces a reader factory that is used for `vast import` and `spawn source` instead of instantiating `source`, `make_source`, `import_command`, etc with every existing reader type.

This change is intended to improve the dev experience during modify -> compile -> run cycles by speeding up incremental builds.

A local test simulating a change to `source.hpp` gives the following times for me:

Before:
```
$ touch libvast/vast/system/source.hpp
$ time cmake --build ~/t/build/vast/debug -j 1
[0/2] Re-checking globbed directories...
[19/19] Linking CXX executable bin/example-test

________________________________________________________
Executed in  155.45 secs   fish           external
   usr time  154.25 secs   38.12 millis  154.21 secs
   sys time    8.83 secs    5.00 millis    8.83 secs
```

After:
```
$ touch libvast/vast/system/source.hpp
$ time cmake --build ~/t/build/vast/detemp -j 1
[0/2] Re-checking globbed directories...
[15/15] Linking integration test directory

________________________________________________________
Executed in   83.25 secs   fish           external
   usr time   85.76 secs    0.00 micros   85.76 secs
   sys time    5.49 secs  712.00 micros    5.49 secs
```

### :dart: Review Instructions

Sorry for the single big commit, it is probably easiest to compare the this against the merge base by checking both locally and opening the header of master next to the `cpp` from this branch, i.e. `master/source.hpp` | `detemplate/source.cpp`.

One can probably start by looking at the changes that were done for the `reader` and derived classes, and then go from outside in: `application.cpp` -> `import_command` -> `make_source` -> `source` & `source_common`. `spawn_source` and `node.cpp` are obvious afterwards.

I'm available for a review call is anyone is interested.

### CODE REVIEW

File: libvast/src/format/reader_factory.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2016 The VAST Contributors
```
Comment (dominiklohmann) [inline libvast/src/format/reader_factory.cpp:6]: This is a new file, so the copyright should say 2021.

This applies to all the other files in this PR as well.

File: libvast/vast/system/datagram_source.hpp [hunk 2]
```diff
@@ -22,26 +22,26 @@
...
-#include "vast/table_slice_builder_factory.hpp"
 
 #include <caf/downstream.hpp>
 #include <caf/event_based_actor.hpp>
-#include <caf/expected.hpp>
 #include <caf/io/broker.hpp>
+#include <caf/settings.hpp>
 #include <caf/stateful_actor.hpp>
-#include <caf/stream_source.hpp>
+//#include <caf/stream_source.hpp>
```
Comment (tobim) [inline libvast/vast/system/datagram_source.hpp:34]: remove

**Comment by tobim** [comment]
@elvisoric you might be interested in this. It should remove some of the friction for exploratory changes.

**Comment by elvisoric** [comment]
@tobim Thank you. I'll take a look.

File: libvast/src/system/make_source.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    }(std::move(*reader), slice_size, max_events, std::move(type_registry),
+      std::move(local_schema), std::move(type_filter), std::move(accountant));
+  VAST_ASSERT(src);
+  // Attempt to parse the remainder as an expression.
+  if (!inv.arguments.empty()) {
+    auto expr = parse_expression(inv.arguments.begin(), inv.arguments.end());
+    if (!expr)
+      return expr.error();
+    anon_send(src, std::move(*expr));
+  }
```
Comment (dominiklohmann) [inline libvast/src/system/make_source.cpp:140]: Side note: Didn't we deprecate this?

File: libvast/src/system/source.cpp [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+#include <caf/event_based_actor.hpp>
+#include <caf/settings.hpp>
+#include <caf/stateful_actor.hpp>
+
+#include <chrono>
+
+namespace vast::system {
+
+caf::behavior
+source(caf::stateful_actor<source_state>* self, format::reader_ptr reader,
```
Comment (dominiklohmann) [inline libvast/src/system/source.cpp:36]: You seem to have reverted a fair amount of recent changes to this function. Please take the most recent version of `source.hpp` for the conversion, and then do the same thing again for the datagram source.

**Review by dominiklohmann — CHANGES_REQUESTED** [review]

File: libvast/src/detail/make_io_stream.cpp [hunk 5]
```diff
@@ -14,16 +14,17 @@
 #include "vast/detail/fdostream.hpp"
 #include "vast/detail/posix.hpp"
 #include "vast/error.hpp"
-#include "vast/logger.hpp"
+#include "vast/path.hpp"
 
 #include <caf/config_value.hpp>
 #include <caf/settings.hpp>
 
+#include <fmt/format.h>
```
[the remark this answers is not in the document]
Comment (tobim) [reply libvast/src/detail/make_io_stream.cpp:22]: We only format strings here, so there is no need to add the heavier include.

File: libvast/src/system/make_source.cpp (hunk 3 above)
Comment (tobim) [reply libvast/src/system/make_source.cpp:140]: Still exists the same way in the current `make_source.hpp`.

Comment (dominiklohmann) [reply libvast/src/system/make_source.cpp:140]: I just think that `inv.arguments.size() <= 1` is always true. Anyways, we can always change this some other time.

File: libvast/src/system/source.cpp (hunk 4 above)
Comment (dominiklohmann) [reply libvast/src/system/source.cpp:36]: @tobim f0029f4 is still incomplete and there are still changes missing.

**Review by dominiklohmann — APPROVED** [review]
I made the requested changes myself, they're all contained in the last commit. @tobim if you approve of my changes, please just merge the PR after CI is done.

File: tools/lsvast/CMakeLists.txt [hunk 6]
```diff
@@ -15,7 +15,7 @@ if (PANDOC_FOUND)
   add_custom_target(lsvast-man ALL
     DEPENDS lsvast
     BYPRODUCTS "${CMAKE_CURRENT_BINARY_DIR}/lsvast.1"
-    COMMAND "${ASCIIDOCTOR}" -s -f markdown -t man
+    COMMAND pandoc -s -f markdown -t man
```
Comment (tobim) [inline tools/lsvast/CMakeLists.txt:18]: Oops. It's fine to fix that here.

File: libvast/vast/system/datagram_source.hpp [hunk 7]
```diff
@@ -10,38 +10,18 @@
...
-#include <caf/event_based_actor.hpp>
-#include <caf/expected.hpp>
-#include <caf/io/broker.hpp>
-#include <caf/stateful_actor.hpp>
-#include <caf/stream_source.hpp>
-#include <caf/streambuf.hpp>
-
-#include <unordered_map>
+#include "caf/event_based_actor.hpp"
+#include "caf/io/broker.hpp"
```
Comment (tobim) [inline libvast/vast/system/datagram_source.hpp:17]: It builds fine without `event_based_actor`.
```suggestion
#include <caf/io/broker.hpp>
```

**Review by tobim — COMMENTED** [review]
Thanks for finishing. Consider this approved from my side.

**Pull request merged**